### PR TITLE
Add ExceptionTracebackFilter + Replace logging.error with logging.exception in API routes

### LINKED ISSUE #2921: All API routes should log errors with stack trace
Many routes log errors with just the exception message which makes it hard to understand some errors and where they come from.
This is especially true in situations where the route returns an HTTP exception to the client.

As example from the log:
```
Failed to create workspace model instance: Only decimal digits permitted in "{'address_space': '10" in "{'address_space': '10.1.177.0"
```
Custom dimension data will show:
```json
{"process":"MainProcess","module":"workspaces","fileName":"/api/./api/routes/workspaces.py","lineNumber":"100","level":"ERROR"}
```
But that doesn't help too much since workspace.py:100 is the line where the log was generated in the route

### PR DESCRIPTION
**Pull request by yuvalyaron**
# Resolves #2921 

## What is being addressed

Improved the logging in the API routes.

## How is this addressed

-  Replaced `logging.error` with `logging.exception` calls in API routes wherever it was applicable - this way we can have the traceback of the exception, and not only the exception message.
-  Added ExceptionTracebackFilter - a logger filter that logs again logs that that have exception info.  This is necessary because if a log has exception info (such as logs that were done using logging.exception) Application insights will only show them in 'exceptions' section without showing in the 'traces' section.

(https://microsoft.github.io/AzureTRE/tre-workspace-authors/authoring-workspace-templates/#versioning)

### CODE REVIEW

File: api_app/services/logging.py [hunk 1]
```diff
@@ -43,6 +43,21 @@ def telemetry_processor_callback_function(envelope):
     envelope.tags['ai.application.ver'] = VERSION
```
Comment (yuvalyaron) [inline api_app/services/logging.py:45]: this is the critical part

File: api_app/api/routes/airlock.py [hunk 2]
```diff
@@ -157,7 +157,7 @@ async def create_review_user_resource(
     try:
         workspace_service = await workspace_service_repo.get_workspace_service_by_id(workspace_id=workspace_id, service_id=workspace_service_id)
     except EntityDoesNotExist as e:
-        logging.error(f"Failed to get workspace service {workspace_service_id} for workspace {workspace_id}: {str(e)}")
+        logging.exception("Failed to get workspace service {workspace_service_id} for workspace {workspace_id}")
```
Comment (tamirkamara) [inline api_app/api/routes/airlock.py:160]:
```suggestion
        logging.exception(f"Failed to get workspace service {workspace_service_id} for workspace {workspace_id}")
```

File: api_app/api/routes/resource_helpers.py [hunk 3]
```diff
@@ -151,8 +151,8 @@ async def send_uninstall_message(
             action=RequestAction.UnInstall,
         )
         return operation
-    except Exception as e:
-        logging.error(f"Failed to send {resource_type} resource delete message: {e}")
+    except Exception:
+        logging.exception("Failed to send {resource_type} resource delete message")
```
Comment (tamirkamara) [inline api_app/api/routes/resource_helpers.py:155]:
```suggestion
        logging.exception(f"Failed to send {resource_type} resource delete message")
```

File: api_app/api/routes/resource_helpers.py [hunk 4]
```diff
@@ -201,10 +201,8 @@ async def send_custom_action_message(
             action=custom_action,
         )
         return operation
-    except Exception as e:
-        logging.error(
-            f"Failed to send {resource_type} resource custom action message: {e}"
-        )
+    except Exception:
+        logging.exception("Failed to send {resource_type} resource custom action message")
```
Comment (tamirkamara) [inline api_app/api/routes/resource_helpers.py:205]:
```suggestion
        logging.exception(f"Failed to send {resource_type} resource custom action message")
```

File: api_app/api/routes/shared_services.py [hunk 5]
```diff
@@ -49,13 +49,13 @@ async def create_shared_service(response: Response, shared_service_input: Shared
     try:
         shared_service, resource_template = await shared_services_repo.create_shared_service_item(shared_service_input, user.roles)
     except (ValidationError, ValueError) as e:
-        logging.error(f"Failed create shared service model instance: {e}")
+        logging.exception("Failed create shared service model instance")
         raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))
     except DuplicateEntity as e:
-        logging.error(f"Shared service already exists: {e}")
+        logging.exception("Shared service already exists:")
```
Comment (tamirkamara) [inline api_app/api/routes/shared_services.py:55]:
```suggestion
        logging.exception("Shared service already exists")
```

File: api_app/services/logging.py [hunk 6]
```diff
@@ -55,6 +70,7 @@ def initialize_logging(logging_level: int, correlation_id: str = None) -> loggin
     logger = logging.getLogger()
 
     try:
+        logger.addFilter(ExceptionTracebackFilter())
```
Comment (tamirkamara) [inline api_app/services/logging.py:73]: This this impact only the logs sent to AppInsights or the ones printed to the console too?

File: api_app/api/routes/airlock.py (hunk 2 above)
Comment (yuvalyaron) [reply api_app/api/routes/airlock.py]: thanks!

**Review by tamirkamara — APPROVED** [review]

**Pull request merged**